[Opengauss] The opengauss database does not support two replicas

Hi team,

Terraform opengauss resource and data source only supports three replicas, but some regions support two replicas.
The logic of the dn number of `huaweicloud_gaussdb_opengauss_instance` is hard-coded and fixed to divide by 3.
So, the two replicas of the opengauss database will throw a panic when getting instance information.

```
Stack trace from the terraform-provider-huaweicloud_v1.31.1 plugin:

panic: runtime error: integer divide by zero

goroutine 45 [running]:
github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud.dataSourceOpenGaussInstanceRead(0xc000d3f700, 0x17d6780, 0xc000790b60, 0x22ad030, 0xc000080000)
        github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/data_source_huaweicloud_gaussdb_opengauss_instance.go:313 +0x1f9c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0005c29a0, 0x1a3ccc8, 0xc000275a80, 0xc000d3f700, 0x17d6780, 0xc000790b60, 0x0, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:335 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0xc0005c29a0, 0x1a3ccc8, 0xc000275a80, 0xc0010e4c60, 0x17d6780, 0xc000790b60, 0xc000790b60, 0xc0010e4c60, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:558 +0xfd
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0xc00000c150, 0x1a3ccc8, 0xc000275a80, 0xc0010e48c0, 0xc000275a80, 0x40b965, 0x16d2420)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/grpc_provider.go:1105 +0x4d6
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ReadDataSource(0xc000e38660, 0x1a3cd70, 0xc000275a80, 0xc000900eb0, 0xc000e38660, 0xc000923530, 0xc000e57ba0)
        github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:247 +0xe5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler(0x175f5a0, 0xc000e38660, 0x1a3cd70, 0xc000923530, 0xc0002f92c0, 0x0, 0x1a3cd70, 0xc000923530, 0xc0001229a0, 0x157)
        github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:416 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000ade1c0, 0x1a49198, 0xc000292900, 0xc000906600, 0xc000c429c0, 0x226acd0, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc000ade1c0, 0x1a49198, 0xc000292900, 0xc000906600, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0008fc200, 0xc000ade1c0, 0x1a49198, 0xc000292900, 0xc000906600)
        google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd

Error: The terraform-provider-huaweicloud_v1.31.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

--- a/docs/data-sources/gaussdb_opengauss_instance.md
+++ b/docs/data-sources/gaussdb_opengauss_instance.md
@@ -57,7 +57,10 @@ In addition to all arguments above, the following attributes are exported:
 
 * `sharding_num` - Indicates the sharding num.
 
+* `replica_num` - Indicates the replica num.
+
 * `private_ips` - Indicates the list of private IP address of the nodes.
+* `public_ips` - Indicates the public IP address of the DB instance.
 
 * `volume` - Indicates the volume information. Structure is documented below.
 

--- a/docs/data-sources/gaussdb_opengauss_instances.md
+++ b/docs/data-sources/gaussdb_opengauss_instances.md
@@ -69,7 +69,11 @@ The `instances` block supports:
 
 * `sharding_num` - Indicates the sharding num.
 
+* `replica_num` - Indicates the replica num.
+
 * `private_ips` - Indicates the list of private IP address of the nodes.
+
+* `public_ips` - Indicates the public IP address of the DB instance.
 
 * `volume` - Indicates the volume information. Structure is documented below.
 

--- a/huaweicloud/services/acceptance/gaussdb/data_source_huaweicloud_gaussdb_opengauss_instance_test.go
+++ b/huaweicloud/services/acceptance/gaussdb/data_source_huaweicloud_gaussdb_opengauss_instance_test.go
@@ -14,6 +14,7 @@ import (
 
 func TestAccOpenGaussInstanceDataSource_basic(t *testing.T) {
 	rName := fmt.Sprintf("tf-acc-test-%s", acctest.RandString(5))
+	dataSourceName := "data.huaweicloud_gaussdb_opengauss_instance.test"
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:          func() { acceptance.TestAccPreCheck(t) },
@@ -22,7 +23,30 @@ func TestAccOpenGaussInstanceDataSource_basic(t *testing.T) {
 			{
 				Config: testAccOpenGaussInstanceDataSource_basic(rName),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckOpenGaussInstanceDataSourceID("data.huaweicloud_gaussdb_opengauss_instance.test"),
+					testAccCheckOpenGaussInstanceDataSourceID(dataSourceName),
+					resource.TestCheckResourceAttr(dataSourceName, "sharding_num", "1"),
+					resource.TestCheckResourceAttr(dataSourceName, "coordinator_num", "2"),
+					resource.TestCheckResourceAttr(dataSourceName, "volume.0.size", "40"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccOpenGaussInstanceDataSource_haModeCentralized(t *testing.T) {
+	rName := fmt.Sprintf("tf-acc-test-%s", acctest.RandString(5))
+	dataSourceName := "data.huaweicloud_gaussdb_opengauss_instance.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { acceptance.TestAccPreCheck(t) },
+		ProviderFactories: acceptance.TestAccProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccOpenGaussInstanceDataSource_haModeCentralized(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckOpenGaussInstanceDataSourceID(dataSourceName),
+					resource.TestCheckResourceAttr(dataSourceName, "replica_num", "3"),
+					resource.TestCheckResourceAttr(dataSourceName, "volume.0.size", "40"),
 				),
 			},
 		},
@@ -46,15 +70,15 @@ func testAccCheckOpenGaussInstanceDataSourceID(n string) resource.TestCheckFunc 
 
 func testAccOpenGaussInstanceDataSource_basic(rName string) string {
 	return fmt.Sprintf(`
-%s
+%[1]s
 
 resource "huaweicloud_networking_secgroup" "test" {
-  name        = "%s"
+  name        = "%[2]s"
   description = "terraform security group rule acceptance test"
 }
 
 resource "huaweicloud_gaussdb_opengauss_instance" "test" {
-  name        = "%s"
+  name        = "%[2]s"
   password    = "Test@12345678"
   flavor      = "gaussdb.opengauss.ee.dn.m6.2xlarge.8.in"
   vpc_id      = huaweicloud_vpc.test.id
@@ -83,5 +107,46 @@ data "huaweicloud_gaussdb_opengauss_instance" "test" {
     huaweicloud_gaussdb_opengauss_instance.test,
   ]
 }
-`, testAccVpcConfig_Base(rName), rName, rName)
+`, testAccVpcConfig_Base(rName), rName)
+}
+
+func testAccOpenGaussInstanceDataSource_haModeCentralized(rName string) string {
+	return fmt.Sprintf(`
+%[1]s
+
+resource "huaweicloud_networking_secgroup" "test" {
+  name        = "%[2]s"
+  description = "terraform security group rule acceptance test"
+}
+
+resource "huaweicloud_gaussdb_opengauss_instance" "test" {
+  name        = "%[2]s"
+  password    = "Test@12345678"
+  flavor      = "gaussdb.opengauss.ee.m6.2xlarge.x868.ha"
+  vpc_id      = huaweicloud_vpc.test.id
+  subnet_id   = huaweicloud_vpc_subnet.test.id
+
+  availability_zone = "cn-north-4a,cn-north-4a,cn-north-4a"
+  security_group_id = huaweicloud_networking_secgroup.test.id
+
+  ha {
+    mode             = "centralization_standard"
+    replication_mode = "sync"
+    consistency      = "strong"
+  }
+  volume {
+    type = "ULTRAHIGH"
+    size = 40
+  }
+
+  replica_num = 3
+}
+
+data "huaweicloud_gaussdb_opengauss_instance" "test" {
+  name = huaweicloud_gaussdb_opengauss_instance.test.name
+  depends_on = [
+    huaweicloud_gaussdb_opengauss_instance.test,
+  ]
+}
+`, testAccVpcConfig_Base(rName), rName)
 }

--- a/huaweicloud/services/acceptance/gaussdb/data_source_huaweicloud_gaussdb_opengauss_instances_test.go
+++ b/huaweicloud/services/acceptance/gaussdb/data_source_huaweicloud_gaussdb_opengauss_instances_test.go
@@ -14,6 +14,7 @@ import (
 
 func TestAccOpenGaussInstancesDataSource_basic(t *testing.T) {
 	rName := fmt.Sprintf("tf-acc-test-%s", acctest.RandString(5))
+	dataSourceName := "data.huaweicloud_gaussdb_opengauss_instances.test"
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:          func() { acceptance.TestAccPreCheck(t) },
@@ -22,8 +23,32 @@ func TestAccOpenGaussInstancesDataSource_basic(t *testing.T) {
 			{
 				Config: testAccOpenGaussInstancesDataSource_basic(rName),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckOpenGaussInstancesDataSourceID("data.huaweicloud_gaussdb_opengauss_instances.test"),
-					resource.TestCheckResourceAttr("data.huaweicloud_gaussdb_opengauss_instances.test", "instances.#", "1"),
+					testAccCheckOpenGaussInstancesDataSourceID(dataSourceName),
+					resource.TestCheckResourceAttr(dataSourceName, "instances.#", "1"),
+					resource.TestCheckResourceAttr(dataSourceName, "instances.0.sharding_num", "1"),
+					resource.TestCheckResourceAttr(dataSourceName, "instances.0.coordinator_num", "2"),
+					resource.TestCheckResourceAttr(dataSourceName, "instances.0.volume.0.size", "40"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccOpenGaussInstancesDataSource_haModeCentralized(t *testing.T) {
+	rName := fmt.Sprintf("tf-acc-test-%s", acctest.RandString(5))
+	dataSourceName := "data.huaweicloud_gaussdb_opengauss_instances.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { acceptance.TestAccPreCheck(t) },
+		ProviderFactories: acceptance.TestAccProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccOpenGaussInstancesDataSource_haModeCentralized(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckOpenGaussInstancesDataSourceID(dataSourceName),
+					resource.TestCheckResourceAttr(dataSourceName, "instances.#", "1"),
+					resource.TestCheckResourceAttr(dataSourceName, "instances.0.replica_num", "3"),
+					resource.TestCheckResourceAttr(dataSourceName, "instances.0.volume.0.size", "40"),
 				),
 			},
 		},
@@ -47,10 +72,10 @@ func testAccCheckOpenGaussInstancesDataSourceID(n string) resource.TestCheckFunc
 
 func testAccOpenGaussInstancesDataSource_basic(rName string) string {
 	return fmt.Sprintf(`
-%s
+%[1]s
 
 resource "huaweicloud_networking_secgroup" "test" {
-  name        = "%s"
+  name        = "%[2]s"
   description = "terraform security group rule acceptance test"
 }
 
@@ -63,7 +88,7 @@ resource "huaweicloud_networking_secgroup_rule" "test" {
 }
 
 resource "huaweicloud_gaussdb_opengauss_instance" "test" {
-  name      = "%s"
+  name      = "%[2]s"
   password  = "Test@12345678"
   flavor    = "gaussdb.opengauss.ee.dn.m6.2xlarge.8.in"
   vpc_id    = huaweicloud_vpc.test.id
@@ -84,6 +109,55 @@ resource "huaweicloud_gaussdb_opengauss_instance" "test" {
 
   sharding_num    = 1
   coordinator_num = 2
+}
+
+data "huaweicloud_gaussdb_opengauss_instances" "test" {
+  name = huaweicloud_gaussdb_opengauss_instance.test.name
+  depends_on = [
+    huaweicloud_gaussdb_opengauss_instance.test,
+  ]
+}
+`, testAccVpcConfig_Base(rName), rName)
+}
+
+func testAccOpenGaussInstancesDataSource_haModeCentralized(rName string) string {
+	return fmt.Sprintf(`
+%[1]s
+
+resource "huaweicloud_networking_secgroup" "test" {
+  name        = "%[2]s"
+  description = "terraform security group rule acceptance test"
+}
+
+resource "huaweicloud_networking_secgroup_rule" "test" {
+  security_group_id = huaweicloud_networking_secgroup.test.id
+  direction         = "ingress"
+  ethertype         = "IPv4"
+  protocol          = "tcp"
+  remote_ip_prefix  = "0.0.0.0/0"
+}
+
+resource "huaweicloud_gaussdb_opengauss_instance" "test" {
+  name      = "%[2]s"
+  password  = "Test@12345678"
+  flavor    = "gaussdb.opengauss.ee.m6.2xlarge.x868.ha"
+  vpc_id    = huaweicloud_vpc.test.id
+  subnet_id = huaweicloud_vpc_subnet.test.id
+
+  availability_zone = "cn-north-4a,cn-north-4a,cn-north-4a"
+  security_group_id = huaweicloud_networking_secgroup.test.id
+
+  ha {
+    mode             = "centralization_standard"
+    replication_mode = "sync"
+    consistency      = "strong"
+  }
+  volume {
+    type = "ULTRAHIGH"
+    size = 40
+  }
+
+  replica_num = 3
 }
 
 data "huaweicloud_gaussdb_opengauss_instances" "test" {

--- a/huaweicloud/services/gaussdb/data_source_huaweicloud_gaussdb_opengauss_instance.go
+++ b/huaweicloud/services/gaussdb/data_source_huaweicloud_gaussdb_opengauss_instance.go
@@ -135,6 +135,10 @@ func DataSourceOpenGaussInstance() *schema.Resource {
 				Type:     schema.TypeInt,
 				Computed: true,
 			},
+			"replica_num": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
 			"volume": {
 				Type:     schema.TypeList,
 				Computed: true,
@@ -152,6 +156,13 @@ func DataSourceOpenGaussInstance() *schema.Resource {
 				},
 			},
 			"private_ips": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"public_ips": {
 				Type:     schema.TypeList,
 				Computed: true,
 				Elem: &schema.Schema{
@@ -243,6 +254,7 @@ func dataSourceOpenGaussInstanceRead(d *schema.ResourceData, meta interface{}) e
 	d.Set("port", instance.Port)
 	d.Set("switch_strategy", instance.SwitchStrategy)
 	d.Set("maintenance_window", instance.MaintenanceWindow)
+	d.Set("public_ips", instance.PublicIps)
 
 	if len(instance.PrivateIps) > 0 {
 		private_ips := instance.PrivateIps[0]
@@ -263,8 +275,9 @@ func dataSourceOpenGaussInstanceRead(d *schema.ResourceData, meta interface{}) e
 	d.Set("datastore", dbList)
 
 	// set nodes
-	sharding_num := 0
-	coordinator_num := 0
+	var dnNum int
+	shardingNum := 0
+	coordinatorNum := 0
 	azList := []string{}
 	nodesList := make([]map[string]interface{}, 0, 1)
 	for _, raw := range instance.Nodes {
@@ -278,16 +291,23 @@ func dataSourceOpenGaussInstanceRead(d *schema.ResourceData, meta interface{}) e
 		nodesList = append(nodesList, node)
 		azList = append(azList, raw.AvailabilityZone)
 		if strings.Contains(raw.Name, "_gaussdbv5cn") {
-			coordinator_num += 1
+			coordinatorNum++
 		} else if strings.Contains(raw.Name, "_gaussdbv5dn") {
-			sharding_num += 1
+			shardingNum++
 		}
 	}
-	d.Set("nodes", nodesList)
-	d.Set("coordinator_num", coordinator_num)
 
-	dn_num := sharding_num / 3
-	d.Set("sharding_num", dn_num)
+	if shardingNum > 0 && coordinatorNum > 0 {
+		dnNum = shardingNum / 3
+		d.Set("nodes", nodesList)
+		d.Set("sharding_num", dnNum)
+		d.Set("coordinator_num", coordinatorNum)
+	} else {
+		// If the HA mode is centralized, the HA structure of API response is nil.
+		dnNum = instance.ReplicaNum + 1
+		d.Set("nodes", nodesList)
+		d.Set("replica_num", instance.ReplicaNum)
+	}
 
 	//remove duplicate az
 	azList = utils.RemoveDuplicateElem(azList)
@@ -312,12 +332,10 @@ func dataSourceOpenGaussInstanceRead(d *schema.ResourceData, meta interface{}) e
 	d.Set("ha", haList)
 
 	// set volume
-	volume_size := instance.Volume.Size
-	dn_size := volume_size / dn_num
 	volumeList := make([]map[string]interface{}, 1)
 	volume := map[string]interface{}{
 		"type": instance.Volume.Type,
-		"size": dn_size,
+		"size": instance.Volume.Size / dnNum,
 	}
 	volumeList[0] = volume
 	d.Set("volume", volumeList)

--- a/huaweicloud/services/gaussdb/data_source_huaweicloud_gaussdb_opengauss_instances.go
+++ b/huaweicloud/services/gaussdb/data_source_huaweicloud_gaussdb_opengauss_instances.go
@@ -159,6 +159,10 @@ func DataSourceOpenGaussInstances() *schema.Resource {
 							Type:     schema.TypeInt,
 							Computed: true,
 						},
+						"replica_num": {
+							Type:     schema.TypeInt,
+							Computed: true,
+						},
 						"volume": {
 							Type:     schema.TypeList,
 							Computed: true,
@@ -176,6 +180,13 @@ func DataSourceOpenGaussInstances() *schema.Resource {
 							},
 						},
 						"private_ips": {
+							Type:     schema.TypeList,
+							Computed: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+						"public_ips": {
 							Type:     schema.TypeList,
 							Computed: true,
 							Elem: &schema.Schema{
@@ -261,6 +272,7 @@ func dataSourceOpenGaussInstancesRead(ctx context.Context, d *schema.ResourceDat
 			"port":                  instanceInAll.Port,
 			"switch_strategy":       instanceInAll.SwitchStrategy,
 			"maintenance_window":    instanceInAll.MaintenanceWindow,
+			"public_ips":            instanceInAll.PublicIps,
 		}
 
 		instanceID := instanceInAll.Id
@@ -285,8 +297,9 @@ func dataSourceOpenGaussInstancesRead(ctx context.Context, d *schema.ResourceDat
 		instanceToSet["datastore"] = dbList
 
 		// set nodes
-		sharding_num := 0
-		coordinator_num := 0
+		var dnNum int
+		shardingNum := 0
+		coordinatorNum := 0
 		azList := []string{}
 		nodesList := make([]map[string]interface{}, 0, 1)
 		for _, raw := range instanceInAll.Nodes {
@@ -300,16 +313,23 @@ func dataSourceOpenGaussInstancesRead(ctx context.Context, d *schema.ResourceDat
 			nodesList = append(nodesList, node)
 			azList = append(azList, raw.AvailabilityZone)
 			if strings.Contains(raw.Name, "_gaussdbv5cn") {
-				coordinator_num += 1
+				coordinatorNum++
 			} else if strings.Contains(raw.Name, "_gaussdbv5dn") {
-				sharding_num += 1
+				shardingNum++
 			}
 		}
-		instanceToSet["nodes"] = nodesList
-		instanceToSet["coordinator_num"] = coordinator_num
 
-		dn_num := sharding_num / 3
-		instanceToSet["sharding_num"] = dn_num
+		if shardingNum > 0 && coordinatorNum > 0 {
+			dnNum = shardingNum / 3
+			instanceToSet["nodes"] = nodesList
+			instanceToSet["coordinator_num"] = coordinatorNum
+			instanceToSet["sharding_num"] = dnNum
+		} else {
+			// If the HA mode is centralized, the HA structure of API response is nil.
+			dnNum = instanceInAll.ReplicaNum + 1
+			instanceToSet["nodes"] = nodesList
+			instanceToSet["replica_num"] = instanceInAll.ReplicaNum
+		}
 
 		//remove duplicate az
 		azList = utils.RemoveDuplicateElem(azList)
@@ -334,12 +354,10 @@ func dataSourceOpenGaussInstancesRead(ctx context.Context, d *schema.ResourceDat
 		instanceToSet["ha"] = haList
 
 		// set volume
-		volume_size := instanceInAll.Volume.Size
-		dn_size := volume_size / dn_num
 		volumeList := make([]map[string]interface{}, 1)
 		volume := map[string]interface{}{
 			"type": instanceInAll.Volume.Type,
-			"size": dn_size,
+			"size": instanceInAll.Volume.Size / dnNum,
 		}
 		volumeList[0] = volume
 		instanceToSet["volume"] = volumeList

